File: cchess/lib/core/chess_engine/position.dart
import 'package:equatable/equatable.dart';

/// Coordinate on the 10×9 Xiangqi board.
///
/// Convention used everywhere in this engine:
/// - `row` is 0..9. Row 0 is Black's back rank (top of screen); row 9 is
///   Red's back rank (bottom of screen).
/// - `col` is 0..8 left-to-right from Red's perspective.
class Position extends Equatable implements Comparable<Position> {
  final int row;
  final int col;

  const Position(this.row, this.col);

  bool get isValid => row >= 0 && row <= 9 && col >= 0 && col <= 8;

  /// Whether this square is within the 3×3 palace for the given color.
  bool isInPalace(bool forRed) {
    if (col < 3 || col > 5) return false;
    if (forRed) {
      return row >= 7 && row <= 9;
    } else {
      return row >= 0 && row <= 2;
    }
  }

  /// Whether this square is on the red half of the board (rows 5..9).
  bool get isOnRedSide => row >= 5;

  /// Whether this square is on the black half of the board (rows 0..4).
  bool get isOnBlackSide => row <= 4;

  Position offset(int dRow, int dCol) => Position(row + dRow, col + dCol);

  @override
  int compareTo(Position other) {
    final r = row.compareTo(other.row);
    if (r != 0) return r;
    return col.compareTo(other.col);
  }

  @override
  String toString() => '($row,$col)';

  @override
  List<Object?> get props => [row, col];
}

File: cchess-backend/src/engine/position.ts
// Port of cchess/lib/core/chess_engine/position.dart
//
// row 0..9, col 0..8. row 0 = Black's back rank (top of screen),
// row 9 = Red's back rank (bottom of screen).

export class Position {
  constructor(public readonly row: number, public readonly col: number) {}

  get isValid(): boolean {
    return this.row >= 0 && this.row <= 9 && this.col >= 0 && this.col <= 8;
  }

  /// Whether this square is within the 3×3 palace for the given color.
  isInPalace(forRed: boolean): boolean {
    if (this.col < 3 || this.col > 5) return false;
    if (forRed) return this.row >= 7 && this.row <= 9;
    return this.row >= 0 && this.row <= 2;
  }

  offset(dRow: number, dCol: number): Position {
    return new Position(this.row + dRow, this.col + dCol);
  }

  equals(other: Position): boolean {
    return this.row === other.row && this.col === other.col;
  }

  toString(): string {
    return `(${this.row},${this.col})`;
  }
}
